plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -417,11 +417,92 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
          packed_transaction_ptr  trx;
          next_func_t             next;
       };
+   
       // The queue storing read-only transactions to be executed by read-only threads
-      struct ro_trx_queue_t {
-         std::mutex              mtx;
-         std::deque<ro_trx_t>    queue;
+      class ro_trx_queue_t {
+      public:
+         void push_back(ro_trx_t&& trx) {
+            std::unique_lock<std::mutex> g( mtx );
+            queue.push_back(std::move(trx));
+            notify_waiting(g, false);
+         }
+         
+         void push_front(ro_trx_t&& trx) {
+            std::unique_lock<std::mutex> g( mtx );
+            queue.push_front(std::move(trx));
+            notify_waiting(g, false);
+         }
+
+         bool empty() const {
+            std::lock_guard<std::mutex> g( mtx );
+            return queue.empty();
+         }
+
+         // may wait if the queue is empty, and not all other threads are already waiting.
+         // returns true if a transaction was dequeued and should be executed, or false 
+         // if conditions are met to stop processing transactions. 
+         bool pop_front(ro_trx_t &trx) {
+            std::unique_lock<std::mutex> g( mtx );
+            
+            if (should_exit()) {
+               notify_waiting(g, true);
+               return false;
+            }
+
+            if (queue.empty()) {
+               if (num_waiting + 1 == max_waiting) {
+                  notify_waiting(g, true);
+                  return false;
+               }
+               
+               ++num_waiting;
+               cond.wait(g);
+               --num_waiting;
+               
+               if (queue.empty() || should_exit()) // were we woken up by another thread?
+                  return false;
+            }
+            
+            trx = std::move(queue.front());
+            queue.pop_front();
+            return true;
+         }
+
+         // We exit the read window when either:
+         //    - all threads would be idle
+         //    - or the net_plugin received a block.
+         //    - or we have reached the read_window_deadline
+         void set_exit_criteria(uint32_t num_tasks, std::atomic<bool>* received_block, fc::time_point deadline) {
+            assert(num_tasks > 0 && num_waiting == 0);
+            assert(received_block && *received_block == false);
```

Comment:
I thought we decided not to use `assert`

Reply:
My memory of it is that we didn't reach a conclusion either way.

Reply:
Areg just said that he is OK with this use of `assert`.